src/imitation/algorithms/sqil.py

```diff
@@ -0,0 +1,314 @@
+            tensorboard_log: The log location for tensorboard (if None, no logging).
+            policy_kwargs: Additional arguments to be passed to the policy on creation.
+            verbose: Verbosity level: 0 for no output, 1 for info messages
+                (such as device or wrappers used), 2 for debug messages.
+            seed: Seed for the pseudo random generators.
+            device: Device (cpu, cuda, ...) on which the code should be run.
+                Setting it to auto, the code will be run on the GPU if possible.
+            _init_setup_model: Whether or not to build the network
+                at the creation of the instance.
+
+
+        """
+        self.venv = venv
+
+        super().__init__(demonstrations=demonstrations, custom_logger=custom_logger)
+
+        self.dqn = dqn.DQN(
+            policy=policy,
+            env=venv,
+            learning_rate=learning_rate,
+            buffer_size=buffer_size,
+            learning_starts=learning_starts,
+            batch_size=batch_size,
+            tau=tau,
+            gamma=gamma,
+            train_freq=train_freq,
+            gradient_steps=gradient_steps,
+            replay_buffer_class=replay_buffer_class,
+            replay_buffer_kwargs=replay_buffer_kwargs,
+            optimize_memory_usage=optimize_memory_usage,
+            target_update_interval=target_update_interval,
+            exploration_fraction=exploration_fraction,
+            exploration_initial_eps=exploration_initial_eps,
+            exploration_final_eps=exploration_final_eps,
+            max_grad_norm=max_grad_norm,
+            tensorboard_log=tensorboard_log,
+            policy_kwargs=policy_kwargs,
+            verbose=verbose,
+            seed=seed,
+            device=device,
+            _init_setup_model=_init_setup_model,
+        )
+
+    def set_demonstrations(self, demonstrations: algo_base.AnyTransitions) -> None:
+        # If demonstrations is a list of trajectories,
+        # flatten it into a list of transitions
+        if isinstance(demonstrations, Iterable):
+            (
+                item,
+                demonstrations,
+            ) = util.get_first_iter_element(  # type: ignore[assignment]
+                demonstrations,  # type: ignore[assignment]
+            )
+            if isinstance(item, types.Trajectory):
+                demonstrations = rollout.flatten_trajectories(
+                    demonstrations,  # type: ignore[arg-type]
+                )
+
+        n_samples = len(demonstrations)  # type: ignore[arg-type]
+        self.expert_buffer = buffers.ReplayBuffer(
+            n_samples,
+            self.venv.observation_space,
+            self.venv.action_space,
+            handle_timeout_termination=False,
+        )
+
+        for transition in demonstrations:
+            self.expert_buffer.add(
+                obs=np.array(transition["obs"]),  # type: ignore[index]
```

Comment:
`Transitions` doesn't directly implement `__iter__`, but it inherits from `TransitionsMinimal`, which in turn inherits from/implements `typing.Sequence`.

`typing.Sequence` is essentially the same as `collections.abc.Sequence` (in fact it might be deprecated in favor of the latter in newer versions)

If we have a valid `Sequence`, i.e. something that implements `__getitem__` and `__len__`, then the ABC automatically fills in `__iter__` and a few other methods, see: https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes

If we look at the `__getitem__` of `TransitionsMinimal`, it indexes each individual array and reassembles it as a dictionary or a `TransitionsMinimal`, so we get the right output.

Reply:
Ah, thanks for the explanation, that makes sense!